packages/core/src/download.ts

```diff
@@ -0,0 +1,72 @@
+import os from 'node:os'
+import path from 'node:path'
+import decompress from 'decompress'
+import fs from 'node:fs'
+import { downloadFile } from './util/file'
+import { CKB_CLI_RELEASE_URL, DEFAULT_CKB_CLI_VERSION, DEFAULT_BIN_PATH } from './constants'
+
+export interface CKBCLIDownloadOptions {
+  version?: string
+  dir?: string
+}
+
+export class CKBCLIDownloader {
+  #binName = os.type() === 'Windows_NT' ? 'ckb-cli.exe' : 'ckb-cli'
+  readonly config: Required<CKBCLIDownloadOptions>
+
+  constructor(options?: CKBCLIDownloadOptions) {
+    const version = options?.version || DEFAULT_CKB_CLI_VERSION
+    const saveToPath = options?.dir || DEFAULT_BIN_PATH
+
+    this.config = { dir: saveToPath, version }
+  }
+
+  async downloadIfNotExists(): Promise<string> {
+    if (await this.hasDownloaded()) {
+      return path.join(this.config.dir, this.#binName)
+    }
+
+    const { version, dir } = this.config
+
+    const machine = os.machine()
+    const platform = (() => {
+      const type = os.type()
+
+      if (type === 'Windows_NT') return 'pc-windows-msvc'
+      if (type === 'Darwin') return 'apple-darwin'
+      if (type === 'Linux') return 'unknown-linux-gnu'
+
+      throw new Error('Unsupported OS ' + type)
+    })()
+
+    const extension = (() => {
+      const type = os.type()
+
+      if (type === 'Windows_NT') return 'zip'
+      if (type === 'Darwin') return 'zip'
+      if (type === 'Linux') return 'tar.gz'
+
+      throw new Error('Unsupported OS ' + type)
+    })()
+
+    const zipFilename = `ckb-cli_v${version}_${machine}-${platform}`
+    const url = `${CKB_CLI_RELEASE_URL}/v${version}/${zipFilename}.${extension}`
+    const zipFilePath = path.join(dir, `${zipFilename}.${extension}`)
```

Comment:
A similar logic was found at https://github.com/ckb-js/kuai/pull/319/files#diff-ae462195f23d8854b7f6704f26eacecf18e4eb304780b6f1557fdccaf671ec7fR37
Maybe it can be defined in a shared module cc @Daryl-L

Reply:
OK, I will compare them and give an conclusion.

Reply:
This function will be optimized by PR https://github.com/ckb-js/kuai/pull/319